keps/sig-api-machinery/20190325-unions.md

```diff
@@ -214,24 +238,74 @@ below.
 ### Normalizing on updates
 
 A "normalization process" will run automatically for all creations and
-modifications (either with update or patch). It will happen automatically in order
-to clear fields and update discriminators. This process will run for both
+modifications (either with update or patch). It will happen automatically in
+order to clear fields and update discriminators. This process will run for both
 core-types and CRDs. It will take place before validation. The sent object
 doesn't have to be valid, but fields will be cleared in order to make it valid.
 This process will also happen before fields tracking (server-side apply), so
 changes in discriminator, even if implicit, will be owned by the client making
-the update (and may result in conflicts).
+the update (and may result in conflicts). Note that fields set to null are
+considered unset (and might get cleared).
 
 This process works as follows:
-- If there is a discriminator, and its value has changed, clear all fields but
-  the one specified by the discriminator,
-- If there is no discriminator, or if its value hasn't changed,
-  - if there is exactly one field, set the discriminator when there is one
-    to that value. Otherwise,
-  - compare the fields set before and after. If there is exactly one field
-    added, set the discriminator (if present) to that value, and remove all
-    other fields. if more than one field has been added, leave the process so
-    that validation will fail.
+- If there is more than one field added in the union (compared to the old
+  object), fail with an error,
+- If the discriminator has changed to a non-nil/non-empty value, and there is
+  one new field set but they don't match, return an error,
+- If the discriminator has changed to a non-nil/non-empty value, clear all the
+  fields but the discriminated field,
+- If only one field is left, update a deducible discriminator to that value,
+- if more than one field is left, fail with an error.
```

Comment:
statements like "compared to the old object" and "one *new* field set" make it seem like we need the content of the old object to perform the normalization. I still think it should be deterministic how a given object will be normalized without needing the existing object's data.

I think this order and process would allow that:
1. if a single union member is set and the discriminator is deducible, set the discriminator to match the member
2. if the discriminator is a non-nil/non-empty value, clear union fields that don't match the discriminator
3. if the discriminator doesn't match a set union member, error